include/albatross/src/utils/async_utils.hpp

```diff
@@ -62,6 +62,94 @@ auto async_apply(const std::vector<ValueType> &xs, const ApplyFunction &f) {
+                                      ApplyFunction, ValueType>::value &&
+                                      std::is_same<void, ApplyType>::value,
+                                  int>::type = 0>
+void async_apply_map(const Map<KeyType, ValueType> &xs,
+                     const ApplyFunction &f) {
+  std::vector<std::future<void>> futures;
+  for (const auto &x : xs) {
+    futures.emplace_back(async_safe(f, x.second));
+  }
+  for (auto &f : futures) {
+    f.get();
+  }
+  std::cout << "This was called, void, no key" << std::endl;
+}
+
+template <template <typename...> class Map, typename KeyType,
+          typename ValueType, typename ApplyFunction,
+          typename ApplyType = typename details::value_only_apply_result<
+              ApplyFunction, ValueType>::type,
+          typename std::enable_if<details::is_valid_value_only_apply_function<
+                                      ApplyFunction, ValueType>::value &&
+                                      !std::is_same<void, ApplyType>::value,
+                                  int>::type = 0>
+inline Grouped<KeyType, ApplyType> async_apply_map(
+    const Map<KeyType, ValueType> &xs, const ApplyFunction &f) {
+  Map<KeyType, std::future<ApplyType>> futures;
+  for (const auto &x : xs) {
+    futures[x.first] = async_safe(f, x.second);
+  }
+
+  Map<KeyType, ApplyType> output;
+  for (auto &f : futures) {
+    output[f.first] = f.second.get();
+  }
+  std::cout << "This was called, return, no key" << std::endl;
+  return output;
+}
+
+template <
+    template <typename...> class Map, typename KeyType, typename ValueType,
+    typename ApplyFunction,
+    typename ApplyType = typename details::key_value_apply_result<
+        ApplyFunction, KeyType, ValueType>::type,
+    typename std::enable_if<details::is_valid_key_value_apply_function<
+                                ApplyFunction, KeyType, ValueType>::value &&
+                                std::is_same<void, ApplyType>::value,
+                            int>::type = 0>
+void async_apply_map(const Map<KeyType, ValueType> &xs,
+                     const ApplyFunction &f) {
+  std::vector<std::future<void>> futures;
+  for (const auto &x : xs) {
+    futures.emplace_back(async_safe(f, x.first, x.second));
+  }
+  for (auto &f : futures) {
+    f.get();
+  }
+  std::cout << "This was called, void, with key" << std::endl;
+}
+
+template <
+    template <typename...> class Map, typename KeyType, typename ValueType,
+    typename ApplyFunction,
+    typename ApplyType = typename details::key_value_apply_result<
+        ApplyFunction, KeyType, ValueType>::type,
+    typename std::enable_if<details::is_valid_key_value_apply_function<
+                                ApplyFunction, KeyType, ValueType>::value &&
+                                !std::is_same<void, ApplyType>::value,
+                            int>::type = 0>
+inline Grouped<KeyType, ApplyType> async_apply_map(
```

Comment:
I think this needs to return `Map<>`

Reply:
The non-async versions of this function return `Grouped`, but they also don't use `Map` internally, so I'll change to match those unless I encounter an issue.